src/utils.ts

```diff
@@ -1,11 +1,94 @@
-import { SubmissionMetadata } from "./api";
+import { SubmissionMetadata, TEMPLATES } from "./api";
+import { SchemaDefinition, SlotDefinition } from "./linkml-metamodel";
 
-export const getSubmissionSamples = (submission?: SubmissionMetadata) => {
+export function getSubmissionSamples(submission?: SubmissionMetadata) {
   if (!submission) {
     return [];
   }
-  const environmentalPackageName =
-    submission.metadata_submission.templates[0].replaceAll("-", "_");
-  const sampleDataField = `${environmentalPackageName}_data`;
+  const environmentalPackageName = submission.metadata_submission.packageName;
+  const sampleDataField = TEMPLATES[environmentalPackageName]?.sampleDataSlot;
+  if (!sampleDataField) {
+    return [];
+  }
   return submission.metadata_submission.sampleData[sampleDataField] || [];
-};
+}
+
+export function getSubmissionSample(
+  submission?: SubmissionMetadata,
+  index?: number,
+) {
+  if (!submission || index === undefined) {
+    return undefined;
+  }
+  return getSubmissionSamples(submission).find(
+    (sample) => sample._index === index,
+  );
+}
+
+function compareByRank(a: { rank?: number }, b: { rank?: number }): number {
+  if (a.rank === undefined) {
+    return -1;
+  }
+  if (b.rank === undefined) {
+    return 1;
+  }
+  return a.rank - b.rank;
+}
+
+export interface SlotGroup {
+  name: string;
+  title?: string;
+  rank?: number;
+  slots: SlotDefinition[];
+}
+export function groupClassSlots(
+  schemaDefinition: SchemaDefinition,
+  className: string,
+): SlotGroup[] {
+  const classDefinition = schemaDefinition.classes?.[className];
+  if (!classDefinition) {
+    throw new Error(`Class ${className} not found in schema`);
+  }
```

Comment:
You could pass this function a `SchemaDefinition` that includes classes `A`, `B`, and `C`, but ask the function to group the slots of class `D`. That would cause the lookup in the previous line to return `undefined`. Throwing an error in that case is maybe a bit extreme. If it turns out to be a problem we could also `return []` at that point.